src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java

```diff
@@ -0,0 +1,270 @@
+    // return startAssertion;
+    // }
+
+    /**
+     * @return the authenticated username if authentication was successful, else
+     *         null
+     */
+    public String finishLogin(
+            PublicKeyCredential<AuthenticatorAssertionResponse, ClientAssertionExtensionOutputs> pkc,
+            String sessionId) throws WebAuthnAuthenticationException {
+        try {
+            AssertionRequest assertionRequest = activeAuthentications.get(sessionId);
+            AssertionResult result = rp.finishAssertion(FinishAssertionOptions.builder().request(assertionRequest)
+                    // The PublicKeyCredentialRequestOptions from startAssertion above
+                    .response(pkc).build());
+            if (result.isSuccess() && result.isSignatureCounterValid()) {
+                final WebAuthnUserAccount account = webAuthnUserAccountRepository
+                        .findByUserHandle(result.getUserHandle().getBase64());
+                List<WebAuthnCredential> credentials = webAuthnCredentialsRepository
+                        .findByUserHandle(account.getUserHandle());
+                Optional<WebAuthnCredential> toUpdate = Optional.empty();
+                ByteArray resultCredentialId = result.getCredentialId();
+                for (WebAuthnCredential c : credentials) {
+                    ByteArray cCredentialId = ByteArray.fromBase64(c.getCredentialId());
+                    if (cCredentialId.equals(resultCredentialId)) {
+                        toUpdate = Optional.of(c);
+                    }
+                }
+                if (toUpdate.isEmpty()) {
+                    throw new WebAuthnAuthenticationException(account.getSubject(),
+                            "Could not find the requested credential in the account");
+                }
+                WebAuthnCredential credential = toUpdate.get();
+                credentials.remove(credential);
+                credential.setSignatureCount(result.getSignatureCount());
+                credential.setLastUsedOn(new Date());
+                webAuthnCredentialsRepository.save(credential);
+                final String username = account.getUsername();
+                if (!StringUtils.hasText(username)) {
+                    throw new WebAuthnAuthenticationException(account.getSubject(),
+                            "Could not find the requested credential in the account");
+                }
+                return username;
+            } else {
+                return null;
+            }
+        } catch (Exception e) {
+            throw new WebAuthnAuthenticationException("_",
+                    "Login failed");
+        }
+    }
+
+    UserIdentity generateUserIdentity(String username, String realm, String displayName,
+            Subject subjectOrNull) {
+        final String userDisplayName = displayName != null ? displayName : "";
+        byte[] userHandle = new byte[64];
+        random.nextBytes(userHandle);
```

Comment:
replace user handle manual generation with a key generator, for example SecureStringKeyGenerator

Reply:
Done in 6ef293a3be2772488db81a7aa1150acf40654ae0